dds/DCPS/JsonValueReader.h

```diff
@@ -0,0 +1,569 @@
+    , input_stream_(input_stream)
+  {
+    reader_.IterativeParseInit();
+  }
+
+  bool begin_struct();
+  bool end_struct();
+  bool begin_struct_member(XTypes::MemberId& member_id, const MemberHelper& helper);
+  bool end_struct_member();
+
+  bool begin_union();
+  bool end_union();
+  bool begin_discriminator();
+  bool end_discriminator();
+  bool begin_union_member();
+  bool end_union_member();
+
+  bool begin_array();
+  bool end_array();
+  bool begin_sequence();
+  bool elements_remaining();
+  bool end_sequence();
+  bool begin_element();
+  bool end_element();
+
+  bool read_boolean(ACE_CDR::Boolean& value);
+  bool read_byte(ACE_CDR::Octet& value);
+  bool read_int8(ACE_CDR::Char& value);
+  bool read_uint8(ACE_CDR::Octet& value);
+  bool read_int16(ACE_CDR::Short& value);
+  bool read_uint16(ACE_CDR::UShort& value);
+  bool read_int32(ACE_CDR::Long& value);
+  bool read_uint32(ACE_CDR::ULong& value);
+  bool read_int64(ACE_CDR::LongLong& value);
+  bool read_uint64(ACE_CDR::ULongLong& value);
+  bool read_float32(ACE_CDR::Float& value);
+  bool read_float64(ACE_CDR::Double& value);
+  bool read_float128(ACE_CDR::LongDouble& value);
+  bool read_fixed(OpenDDS::FaceTypes::Fixed& value);
+  bool read_char8(ACE_CDR::Char& value);
+  bool read_char16(ACE_CDR::WChar& value);
+  bool read_string(std::string& value);
+  bool read_wstring(std::wstring& value);
+  bool read_long_enum(ACE_CDR::Long& value, const EnumHelper& helper);
+
+  bool Null() { token_type_ = kNull; return true; }
+  bool Bool(bool b) { token_type_ = kBool; bool_value_ = b; return true; }
+  bool Int(int i) { token_type_ = kInt; int_value_ = i; return true; }
+  bool Uint(unsigned i) { token_type_ = kUint; uint_value_ = i; return true; }
+  bool Int64(int64_t i) { token_type_ = kInt64; int64_value_ = i; return true; }
+  bool Uint64(uint64_t i) { token_type_ = kUint64; uint64_value_ = i; return true; }
+  bool Double(double d) { token_type_ = kDouble; double_value_ = d; return true; }
+  bool RawNumber(const Ch* str, rapidjson::SizeType length, bool copy) { token_type_ = kRawNumber; return true; }
+  bool String(const Ch* str, rapidjson::SizeType length, bool copy) { token_type_ = kString; string_value_ = str; return true; }
+  bool StartObject() { token_type_ = kStartObject; return true; }
+  bool Key(const Ch* str, rapidjson::SizeType length, bool copy) { token_type_ = kKey; key_value_ = str; return true; }
+  bool EndObject(rapidjson::SizeType /* memberCount */) { token_type_ = kEndObject; return true; }
+  bool StartArray() { token_type_ = kStartArray; return true; }
+  bool EndArray(rapidjson::SizeType elementCount) { token_type_ = kEndArray; return true; }
+
+private:
+  enum TokenType {
+    kUnknown,
+    kNull,
+    kBool,
+    kInt,
+    kUint,
+    kInt64,
+    kUint64,
+    kDouble,
+    kRawNumber,
+    kString,
+    kStartObject,
+    kKey,
+    kEndObject,
+    kStartArray,
+    kEndArray,
+    kError,
+    kEnd
+  };
+
+  void peek() {
```

Comment:
Should this function return the token's type?  That may simplify the code that calls it.

Reply:
I don't think so.  You are either going to do
`const TokenType x = peek();` which is unnecessary since the logical return type of peek is the member `token_type_` or you are going to do
`if (peek() == X)` which only gains a line a put a modifying statement in an if which I tend to not like.

Reply:
I was thinking of this usage:
```
  switch (peek()) {
  case kInt: 
    value = int_value_;
    return consume(kInt);
  case kUint:
    value = uint_value_;
    return consume(kUint);
  }
```

Reply:
That makes sense.